Apply schema filters everywhere

gitfool opened the pull request:
Following up from #776; make `ISchemaFilter` available for all types.

Unit tests use a `ISchemaFilter` to demonstrate this for all primitives (including enums), composites and arrays, including with nullable types, used in parameters and request / response types.

This could be made available in a separate AutoRest extension as follows:
```c#
public class XmsTypeDotNetSchemaFilter : ISchemaFilter
{
    public void Apply(Schema schema, SchemaRegistry schemaRegistry, Type type)
    {
        if (type.IsArray && type.GetElementType() != typeof(byte)) return; // Special case

        var nullableUnderlyingType = Nullable.GetUnderlyingType(type);

        schema.vendorExtensions.Add("x-type-dotnet", (nullableUnderlyingType ?? type).FullName);
        schema.vendorExtensions.Add("x-nullable", nullableUnderlyingType != null || !type.IsValueType);
    }
}
```
Also, x-ms-enum support could now be applied using an `ISchemaFilter` as follows:
```c#
public class XmsEnumSchemaFilter : ISchemaFilter
{
    public void Apply(Schema schema, SchemaRegistry schemaRegistry, Type type)
    {
        type = Nullable.GetUnderlyingType(type) ?? type;

        if (type.IsEnum)
        {
            schema.vendorExtensions.Add("x-ms-enum", new { name = type.Name });
        }
    }
}
```
And "required" properties could be derived from non-nullable properties as follows:
```c#
public static class SchemaExtensions
{
    public static bool IsNullable(this Schema schema)
    {
        object value;
        return (schema.vendorExtensions != null) && schema.vendorExtensions.TryGetValue("x-nullable", out value) && (value is bool) && (bool)value;
    }
}

public class RequiredSchemaFilter : ISchemaFilter
{
    public void Apply(Schema schema, SchemaRegistry schemaRegistry, Type type)
    {
        var required = schema.properties?.Where(x => !x.Value.IsNullable())
            .Select(x => x.Key)
            .ToArray();

        if (required?.Length > 0)
        {
            if (schema.required == null)
            {
                schema.required = new List<string>();
            }
            foreach (var property in required)
            {
                schema.required.Add(property);
            }
        }
    }
}
```

gitfool commented:
@domaindrivendev Should `MapType` be deprecated or removed as part of these changes?

domaindrivendev commented:
I think we should still support it as they serve different use cases. The filters are typically applied across the board so you can add additional metadata to all schemas based on some provided logic. MapType on the other hand is simply a wholesale override of the Schema for a specific type. For example, if you had a C# class that, for whatever reason, is serialized to a string in the API.

Does that seem reasonable to you?

gitfool commented:
I'm not bothered either way, but it is the odd duck that doesn't get filtered; if it's kept then maybe it should be filtered too, for consistency?

Otherwise I think I'm done. :wink:

gitfool commented:
@domaindrivendev It's been a while. In the meantime I've realized that I'd much prefer for mapped types to be filtered too. I pushed a separate commit to show the difference including a unit test. If you're in agreement then I can squash the commits.

domaindrivendev commented:
I'm seeing a lot of red tests when I merge this in - did you get green for the entire suite?

gitfool commented:
Check again just now. All green:

![image](https://cloud.githubusercontent.com/assets/750121/16722003/bac2c244-4797-11e6-8ffa-4419d9662b2d.png)

domaindrivendev commented on Swashbuckle.Core/Swagger/SchemaRegistry.cs:
```diff
@@ -81,46 +81,46 @@ public Schema GetOrRegister(Type type)
 
```
Having multiple calls to __FilterSchema__ scattered throughout this class bothers me a little. However, it possibly points to more of a smell with the existing implementation of SchemaRegistry than the code you're adding. I think it makes sense for now but it's starting to make me think there should be a single private method that all Schema's are generated through (delegating to more specific methods where applicable) and this should be the place where the filters get applied. Perhaps a refactor for another time ...

gitfool replied:
Agreed. This crossed my mind too, but given the current code structure it seemed reasonable.

domaindrivendev commented on Swashbuckle.Tests/Swagger/SchemaTests.cs:
```diff
@@ -301,16 +302,987 @@ public void It_exposes_config_to_map_a_type_to_an_explicit_schema()
+            };
+            Assert.AreEqual(JObject.FromObject(expectedResponse).ToString(), response.ToString());
+        }
+
```
I think the number of TestCases here is overkill and bloats the test codebase. Wouldn't it be sufficient to just have 1 TestCase for a mapped type, 1 TestCase for an enum, 1 TestCase for a primitive etc?

domaindrivendev commented on Swashbuckle.Tests/Swagger/SchemaTests.cs:
```diff
@@ -301,16 +302,987 @@ public void It_exposes_config_to_map_a_type_to_an_explicit_schema()
+                })
+            });
+            Assert.AreEqual(expectedProps.ToString(), compositeProps.ToString());
         }
```
Put another way - this feature doesn't warrant over 800 lines of test code!

gitfool commented:
I rebased and squashed since I was behind upstream/master. This revealed the breaking unit tests in XmlCommentsTests. I then fixed these tests which was due to [ApplyXmlTypeComments.cs#L38](https://github.com/gitfool/Swashbuckle/blob/a978120ffdad66fc8c47a1f09dd88c683e375d66/Swashbuckle.Core/Swagger/XmlComments/ApplyXmlTypeComments.cs#L38) not checking for null properties, which otherwise failed because it's now called from [here](https://github.com/gitfool/Swashbuckle/blob/a978120ffdad66fc8c47a1f09dd88c683e375d66/Swashbuckle.Core/Swagger/SchemaRegistry.cs#L109) due to the refactor with _everything_ going through FilterSchema.

As for the tests, I'm happy to remove them, but the idea was to verify the schema filter was being used "for all primitives (including enums), composites and arrays, including with nullable types, used in parameters and request / response types", since this was not the case before.

gitfool commented:
@domaindrivendev I removed the composite related tests, but left the other tests since they're fairly compact using TestCase attributes. Let me know if that sufficiently addresses your concerns. If so, then I'll squash again before you would merge.

domaindrivendev commented:
@gitfool - the TestCase syntax may be more compact but the comprehensive set of cases for every primitive is redundant. At the end of the day, they're all just testing the same line of code:

    if (jsonContract is JsonPrimitiveContract)
        return FilterSchema(CreatePrimitiveSchema((JsonPrimitiveContract)jsonContract), jsonContract);

I would also say that the tests for this feature should only be concerned with whether or not a filter get's applied. So, I'd prefer them to focus on this one thing instead of testing all the other stuff like "type", "format", parameter values and even the actual logic inside the filter.

I'm happy to accept this as is because we may be running into diverging testing philosophies and I don't want that to get in the way of a good addition to this project. However, I may look to simplify the tests at some stage further down the road, possibly when I look at refactoring the internals of SchemaRegistry.

gitfool commented:
@domaindrivendev I think we're on the same page. I've squashed the commits again so you can at least cleanly merge this first step _now_.

Ultimately the reason for the tests is I wanted to cover the logic for the filter itself, given all the possible inputs and the odd special case. So they have value, just not in Swashbuckle Core, so I think they should be moved to be tests associated with a new Swashbuckle.AutoRest (or similar) project.

gitfool commented:
Rebased with latest changes from upstream/master.

gitfool commented:
@domaindrivendev Any chance we can merge this sooner rather than later? This is a key change that enables forward progress in other areas...

domaindrivendev commented:
Ready to merge but hesitating because it occurred to me this may be a breaking change for some consumers. Any custom Schema filters our there are now going to be invoked for all Schema's, not just "object" Schema's. This means the "Properties" property may be null. The semantics have changed.

I'm OK with introducing breaking changes to move the project forward but generally follow semantic versioning, restricting this to major version releases only. However, the next major version (6.0.0) is already deep into beta. It's a "massive" change (actually a rewrite), targeting the next-gen framework ASP.NET Core. So much so, that the source code for that currently resides in a different repo - https://github.com/domaindrivendev/Ahoy.

Anyway, this leaves us with 3 options as I see it:

1) Target this change to the next-gen version (i.e. create the PR in the Ahoy repo)
- the downside here is that consumers don't get it unless they make the massive jump to ASP.NET Core and SB 6.0.0.

2) Break semantic versioning and release with SB 5.4.0
- the downside here is we may break some custom Schema's out there if they're making the assumption that "Properties" is always non-null.

3) Add a temporary config. flag - "ApplyFiltersToAllSchemas" and disable by default to ensure backwards compat.

Thoughts?

gitfool commented:
Option (3) sounds like the way to go for this repo and 5.4.0.

I should also open a PR to get it into 6.0.0 in the other repo. That way you can release 5.4.0 before 6.0.0 is final, and 6.0.0 will have the same functionality upon release; although we should either remove the config option from 6.0.0 or change the default to enable by default.

domaindrivendev commented:
Sounds good - can you update this PR. I'm in the process of porting to the other repo (Ahoy) so don't worry about that. In 6.0.0, I'm going to remove the config flag

domaindrivendev commented:
Will you annotate the config method with an ObsoleteAttribue explaining that it will be deprecated in 6.0.0

gitfool commented:
@domaindrivendev Have a look at the changes. To keep the code clean I added a default parameter to `FilterSchema` which is only overridden for the schema returned from `CreateObjectSchema`. Hopefully this makes the original behaviour very explicit and easy to see.

domaindrivendev commented:
Thanks dude - made a minor change to the config option logic. Thanks for your patience, onto step 2 :)
